### PR TITLE
[WIP] Support transpose in Python

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Support
- built-in function "transpose"

### CODE REVIEW

File: python/pyspark/sql/functions/builtin.py [hunk 1]
```diff
@@ -5223,6 +5224,72 @@ def broadcast(df: DataFrame) -> DataFrame:
     return DataFrame(cast(JVMView, sc._jvm).functions.broadcast(df._jdf), df.sparkSession)
 
 
+def transpose(
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:5227]: nit, I guess this should be a `dataframe` method? 
```
dataframe.transpose(...)
```

Comment (xinrong-meng) [reply python/pyspark/sql/functions/builtin.py:5227]: Yes, this PR just acts like a prototype to explore the feasibility to resolve "transpose" to "pivot" in analyzer. We have milestones defined [here](https://docs.google.com/document/d/1fWvyU1IeNi7O_P6htC99jfGBD68ted1vEAhG4zb5goM/edit#heading=h.q1f9ayv5c32y).

File: python/pyspark/sql/functions/builtin.py [hunk 2]
```diff
@@ -5223,6 +5224,72 @@ def broadcast(df: DataFrame) -> DataFrame:
...
+    pivot_column_name = "__pivot_column"
+    if column_names is None:
+        column_names = []
+    if order_by is None:
+        order_by = col(pivot_column_name).asc()
+    if column_names:
+        temp_df = temp_df.withColumn(
+            pivot_column_name, concat_ws("_", *[col(name) for name in column_names])
+        )
+        pivot_values = [
```
Comment (xinrong-meng) [inline python/pyspark/sql/functions/builtin.py:5264]: "distinct", "orderBy" and "collect" should not be "executed" in Analyzer, and should be delayed to Physical Planning, is my understanding correct?
Please let me know if the question is not clear enough
@ueshin @zhengruifeng thanks!

**Comment by xinrong-meng** [comment]
```
Logical plan before analysis:
'Transpose ['year, 'sales], variable, 1
+- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).course, true, false, true) AS course#4, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).year AS year#5, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).earnings AS earnings#6]
   +- ExternalRDD [obj#3]
```

```
Logical plan after analysis:
'Pivot List(variable#16), __pivot_column#17: string, [first(value#18) ignore nulls]
+- 'Unpivot List(__pivot_column#15), List(List(course#4), List(earnings#6)), variable, [value]
   +- 'Project [course#4, year#5, earnings#6, concat_ws(_, year#5, 'sales) AS __pivot_column#14]
      +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).course, true, false, true) AS course#4, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).year AS year#5, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).earnings AS earnings#6]
         +- ExternalRDD [obj#3]

```

**Comment by xinrong-meng** [comment]
Logical plan before analysis:
```
'Transpose [dotNET, Java], IntegerType
+- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).course, true, false, true) AS course#4, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).year AS year#5, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).earnings AS earnings#6]
   +- ExternalRDD [obj#3]
```
Logical plan after analysis:
```
Project [key#16, __pivot_first(DISTINCT value) AS `first(DISTINCT value)`#26[0] AS dotNET#27, __pivot_first(DISTINCT value) AS `first(DISTINCT value)`#26[1] AS Java#28]
+- Aggregate [key#16], [key#16, pivotfirst(course#4, first(DISTINCT value)#20, dotNET, Java, 0, 0) AS __pivot_first(DISTINCT value) AS `first(DISTINCT value)`#26]
   +- !Aggregate [key#16, course#4], [key#16, course#4, first(distinct value#14, true) AS first(DISTINCT value)#20]
      +- Expand [[course#4, year, cast(year#5 as double)], [course#4, earnings, earnings#6]], [course#4, key#18, value#19]
         +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).course, true, false, true) AS course#4, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).year AS year#5, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$CourseSales, true])).earnings AS earnings#6]
            +- ExternalRDD [obj#3]
```

**Pull request closed without merging**